File: runtime/ruby/units/wasi/fd_datasync.rb
def wasi_fd_datasync(fd)
  io = @fds[fd]
  return ERRNO_BADF unless io.is_a?(IO)
  begin
    io.fdatasync
  rescue NotImplementedError
    io.fsync # not available on all platforms (e.g. macOS)
  end
  ERRNO_SUCCESS
rescue SystemCallError
  ERRNO_IO
end

File: runtime/python/units/wasi/fd_datasync.py
def wasi_fd_datasync(self, fd):
    io = self.fds.get(fd)
    if io is None or isinstance(io, self.WasiDir):
        return self.ERRNO_BADF
    try:
        if hasattr(os, "fdatasync"):
            os.fdatasync(io.fileno())
        else:
            os.fsync(io.fileno())  # not available on all platforms (macOS)
    except OSError:
        return self.ERRNO_IO
    return self.ERRNO_SUCCESS
